streams/shared/src/main/scala/scalaz/stream/ZSink.scala

```diff
@@ -133,45 +175,63 @@ trait ZSink[-R, +E, +A0, -A, +B] { self =>
       def extract(state: State) = self.extract(state)
     }
 
-  final def filterNot[A1 <: A](f: A1 => Boolean): ZSink[R, E, A0, A1, B] =
-    filter(a => !f(a))
-
-  final def filterNotM[E1 >: E, A1 <: A](f: A1 => IO[E1, Boolean]): ZSink[R, E1, A0, A1, B] =
-    filterM(a => f(a).map(!_))
-
-  final def contramapM[R1 <: R, E1 >: E, C](f: C => ZIO[R1, E1, A]): ZSink[R1, E1, A0, C, B] =
-    new ZSink[R1, E1, A0, C, B] {
+  /**
+   * Effectfully filters the inputs fed to this sink.
+   */
+  final def filterM[R1 <: R, E1 >: E, A1 <: A](f: A1 => IO[E1, Boolean]): ZSink[R1, E1, A0, A1, B] =
+    new ZSink[R1, E1, A0, A1, B] {
       type State = self.State
 
       val initial = self.initial
 
-      def step(state: State, c: C) = f(c).flatMap(self.step(state, _))
+      def step(state: State, a: A1) =
+        f(a).flatMap(
+          b =>
+            if (b) self.step(state, a)
+            else IO.succeed(Step.more(state))
+        )
 
-      def extract(state: State): ZIO[R1, E1, B] = self.extract(state)
+      def extract(state: State) = self.extract(state)
     }
 
-  def contramap[C](f: C => A): ZSink[R, E, A0, C, B] =
-    new ZSink[R, E, A0, C, B] {
-      type State = self.State
-
-      val initial = self.initial
+  /**
+   * Filters this sink by the specified predicate, dropping all elements for
+   * which the predicate evaluates to true.
+   */
+  final def filterNot[A1 <: A](f: A1 => Boolean): ZSink[R, E, A0, A1, B] =
+    filter(a => !f(a))
 
-      def step(state: State, c: C) = self.step(state, f(c))
+  /**
+   * Effectfully ilters this sink by the specified predicate, dropping all elements for
+   * which the predicate evaluates to true.
+   */
+  final def filterNotM[E1 >: E, A1 <: A](f: A1 => IO[E1, Boolean]): ZSink[R, E1, A0, A1, B] =
+    filterM(a => f(a).map(!_))
 
-      def extract(state: State): ZIO[R, E, B] = self.extract(state)
-    }
+  /**
+   * Ignites the sink by producing the initial step.
```

Comment:
Using `ZSchedule` as inspiration...
```suggestion
   * The initial state of the sink.
```